### PR TITLE
feat: ACH transfers

### PR DESCRIPTION
**Pull request by Sangamesh26**
## Type of Change

- [x] New feature

## Description

ACH (Automated Clearing House) transfer is a US based bank transfer method that enables electronic bank-to-bank money transfers through the ACH Network run by Nacha.

Steps followed for ACH transfers:
 1. After getting payments/confirm call for ACH bank transfer, stripe's source endpoint is called. We get the account_no, routing_no , swift_code etc. This response from stripe's endpoint is shown in the Hyperswitch's confirrm call response.
 
 ```
 "ach_credit_transfer": {
        "account_number": "test_4e28a1d0114d",
        "bank_name": "TEST BANK",
        "routing_number": "110000000",
        "swift_code": "TSTEZ122",
        "amount_received": 0,
        "amount_charged": 0
    }
```
2.  Front-end needs to instruct the customer to transfer funds from their bank to the above routing number and account number
3. Once the customer transfers money from their account and the transfer is settled (may take up to 5 days), the above source becomes chargeable and Stripe sends the **source.chargeable** and **source.transaction.created** webhook events.
4. Custom webhook event flow is implemented for ACH transfer, like after getting the above mentioned webhook events, we call **customer** endpoint and get the id and consume it in **charges** call.
5. The charge immediately succeeds but in case of some unexpected error, **charge.succeeded** webhook event from Stripe is received to confirm the payment success.
6. After the successful charge, webhook is triggered to merchant.

## Motivation and Context

Bank Transfers are a popular way of transmitting money between different bank accounts and they are popular in US, EU and few Asian and LATAM countries. They are primarily used by businesses for accepting large payments from other businesses. Bank transfers are also used by consumers in certain countries as a preferred method for transferring money to others and while transacting online.
So, ACH bank transfer will be added as a bank transfer feature

## How did you test it?

image
After this, if we get the source chargeable or source transaction created, we call customer and charges api of stripe and trigger the merchant webhook event.

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed submitted code

### CODE REVIEW

**Review by jagan-jaya — CHANGES_REQUESTED** [review]

File: crates/api_models/src/payments.rs [hunk 1]
```diff
@@ -804,6 +823,27 @@ pub struct NextAction {
...
+
+#[derive(Clone, Debug, Eq, PartialEq, serde::Serialize, serde::Deserialize)]
+pub struct NextStepsRequirements {
+    pub ach_credit_transfer: AchTransfer,
+    pub receiver: ReceiverDetails,
+}
+
+#[derive(Clone, Debug, Eq, PartialEq, serde::Serialize, serde::Deserialize)]
+pub struct AchTransfer {
+    pub account_number: String,
```
Comment (jarnura) [inline crates/api_models/src/payments.rs:837]: Make field as secret for account_number, routing_number, swift_code.

**Review by jarnura — COMMENTED** [review]
Add comments for all the field added in the api_models

File: crates/api_models/src/webhooks.rs [hunk 2]
```diff
@@ -52,6 +56,9 @@ impl From<IncomingWebhookEvent> for WebhookFlow {
             IncomingWebhookEvent::DisputeWon => Self::Dispute,
             IncomingWebhookEvent::DisputeLost => Self::Dispute,
             IncomingWebhookEvent::EndpointVerification => Self::ReturnResponse,
+            IncomingWebhookEvent::SourceChargeable => Self::Custom,
```
Comment (Abhicodes-crypto) [inline crates/api_models/src/webhooks.rs:59]:
```suggestion
            IncomingWebhookEvent::SourceChargeable
            | IncomingWebhookEvent::SourceTransactionCreated => Self::Custom
            
```

Comment (Abhicodes-crypto) [reply crates/api_models/src/webhooks.rs:59]: Also you can make a similar change for other branches in this match.

**Review by jarnura — CHANGES_REQUESTED** [review]

File: crates/router/src/db/payment_attempt.rs [hunk 3]
```diff
@@ -513,6 +552,32 @@ mod storage {
...
+                            add_preprocessing_id_to_reverse_lookup(
+                                self,
+                                key.as_str(),
+                                this.merchant_id.as_str(),
+                                updated_attempt.attempt_id.as_str(),
+                                preprocessing_id.as_str(),
+                            )
+                            .await?;
+                        }
+                        (Some(old_preprocessing_id), Some(preprocessing_id)) => {
```
Comment (jarnura) [inline crates/router/src/db/payment_attempt.rs:566]: Throw error in this case

File: migrations/2023-04-20-162755_add_preprocessing_step_id_payment_attempt/up.sql [hunk 4]
```diff
@@ -0,0 +1,2 @@
+-- Your SQL goes here
+ALTER TABLE payment_attempt ADD COLUMN preprocessing_step_id VARCHAR DEFAULT NULL;
```
Comment (jarnura) [inline migrations/2023-04-20-162755_add_preprocessing_step_id_payment_attempt/up.sql:2]: Add Indexing for this field.

File: crates/router/src/types.rs [hunk 5]
```diff
@@ -205,6 +224,13 @@ pub struct CompleteAuthorizeData {
     pub browser_info: Option<BrowserInformation>,
     pub connector_transaction_id: Option<String>,
     pub connector_meta: Option<serde_json::Value>,
+    pub customer_id: Option<String>,
```
Comment (jarnura) [inline crates/router/src/types.rs:227]: This is not required

File: crates/router/src/core/webhooks.rs [hunk 6]
```diff
@@ -358,6 +366,86 @@ async fn disputes_incoming_webhook_flow<W: api::OutgoingWebhookType>(
...
+    source_verified: bool,
+) -> CustomResult<(), errors::WebhooksFlowError> {
+    let response = if source_verified {
+        let payment_attempt = get_payment_attempt_from_object_reference_id(
+            &state,
+            webhook_details.object_reference_id,
+            &merchant_account,
+        )
+        .await?;
+        let request = api::PaymentsRequest {
```
Comment (jarnura) [inline crates/router/src/core/webhooks.rs:382]: Move logic to a separate function.

File: crates/router/src/core/payments/transformers.rs [hunk 7]
```diff
@@ -267,16 +270,50 @@ where
...
+                                            "Failed to parse the Value to NextRequirements struct",
+                                        )
+                                })
+                                .transpose()?;
+                        bank_transfer_next_steps
+                    } else {
+                        None
+                    };
+
+                if payment_intent.status == enums::IntentStatus::RequiresCustomerAction
```
Comment (jarnura) [inline crates/router/src/core/payments/transformers.rs:298]: The condition here is `AND` or `OR`?

File: crates/router/src/core/payments/operations/payment_status.rs [hunk 8]
```diff
@@ -356,6 +356,23 @@ pub async fn get_payment_intent_payment_attempt(
                     )
                     .await?;
             }
+            api_models::payments::PaymentIdType::PreprocessingId(ref id) => {
```
Comment (jarnura) [inline crates/router/src/core/payments/operations/payment_status.rs:359]: Validate this is required or not?

**Review by jarnura — CHANGES_REQUESTED** [review]

File: crates/router/src/core/payments/operations/payment_status.rs (hunk 8 above)
Comment (Sangamesh26) [reply crates/router/src/core/payments/operations/payment_status.rs:359]: This is required,we have added PreprocessingId variant in PaymentIdType, this function expects  to return payment_intent and payment_attempt.

File: crates/router/src/core/payments/transformers.rs (hunk 7 above)
Comment (Sangamesh26) [reply crates/router/src/core/payments/transformers.rs:298]: The condition should be OR, the use case is :
If bank_transfer_next_steps is some, then whatever the intent status may be, we go with bank_transfer_details, if bank_transfer is none and if intent_status is requires_customer_action, we go with RedirectToUrl

File: crates/router/src/core/webhooks.rs (hunk 6 above)
Comment (Sangamesh26) [reply crates/router/src/core/webhooks.rs:382]: This is just calling a function to get payment_attempt and request is formed, there is no specific logic, so we can keep this code here itself.

**Review by jarnura — CHANGES_REQUESTED** [review]

**Review by jarnura — DISMISSED** [review]

File: crates/router/src/connector/stripe/transformers.rs [hunk 9]
```diff
@@ -1003,6 +1058,28 @@ pub struct PaymentIntentResponse {
     pub last_payment_error: Option<ErrorDetails>,
 }
 
+#[derive(Clone, Debug, Default, Eq, PartialEq, Deserialize, Serialize)]
```
[the remark this answers is not in the document]
Comment (Sangamesh26) [reply crates/router/src/connector/stripe/transformers.rs:1290]: Both serialization and Deserialization is required for this struct

File: crates/router/src/core/payments.rs [hunk 10]
```diff
@@ -664,6 +674,38 @@ where
...
+    should_continue_payment: bool,
+) -> RouterResult<(types::RouterData<F, Req, Res>, bool)>
+where
+    F: Send + Clone + Sync,
+    Req: Send + Sync,
+    types::RouterData<F, Req, Res>: Feature<F, Req> + Send,
+    dyn api::Connector: services::api::ConnectorIntegration<F, Req, types::PaymentsResponseData>,
+{
+    let router_data_and_should_continue_payment = match payment_data.payment_method_data.clone() {
+        Some(api_models::payments::PaymentMethodData::BankTransfer(data)) => match data.deref() {
```
[the remark this answers is not in the document]
Comment (Sangamesh26) [reply crates/router/src/core/payments.rs:691]: Added a TODO for it, in future if we encounter any connector not requiring preprocessing steps for ACH transfer, we can add a check for it.I feel, most of the connectors require it(not sure though)

File: crates/router/src/core/payments/transformers.rs [hunk 11]
```diff
@@ -267,16 +270,50 @@ where
...
-                        next_action_type: api::NextActionType::RedirectToUrl,
+                        next_action_type,
                         redirect_to_url: Some(helpers::create_startpay_url(
                             server,
                             &payment_attempt,
                             &payment_intent,
                         )),
-                    })
-                }
+                        bank_transfer_steps_and_charges_details: bank_transfer_next_steps,
```
[the remark this answers is not in the document]
Comment (Sangamesh26) [reply crates/router/src/core/payments/transformers.rs:317]: Added a TODO for it

**Review by Narayanbhat166 — APPROVED** [review]

**Review by jagan-jaya — DISMISSED** [review]
Approving only my PR comments

**Review by Narayanbhat166 — APPROVED** [review]

**Review by jarnura — APPROVED** [review]

**Review by Narayanbhat166 — APPROVED** [review]

**Pull request merged**